warehouse/attestations/_core.py

```diff
@@ -0,0 +1,96 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import hashlib
+import tempfile
+import typing
+
+from pathlib import Path
+
+import sentry_sdk
+
+from pypi_attestations import (
+    Attestation,
+    AttestationBundle,
+    GitHubPublisher,
+    GitLabPublisher,
+    Provenance,
+    Publisher,
+)
+
+from warehouse.attestations.errors import UnsupportedPublisherError
+from warehouse.oidc.models import (
+    GitHubPublisher as GitHubOIDCPublisher,
+    GitLabPublisher as GitLabOIDCPublisher,
+    OIDCPublisher,
+)
+from warehouse.packaging import File, IFileStorage
+
+
+def get_provenance_digest(request, file: File) -> str | None:
+    """Returns the sha256 digest of the provenance file for the release."""
+    if not file.attestations or not file.publisher_url:
```

Comment:
Possibly a silly question, but why do we need the `file.publisher_url` check here? IIUC `file.attestations` should imply `file.publisher_url`, right?

Reply:
Not a silly question at all.
That check was a bit too cautious and did not improve anything - I removed it.